x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -368,80 +446,117 @@ export class ExecutionHandler<
-        (this.skippedAlerts[alertId] && this.skippedAlerts[alertId].reason !== Reasons.MUTED)
-      ) {
-        this.logger.debug(
-          `skipping scheduling of actions for '${alertId}' in rule ${this.ruleLabel}: rule is muted`
-        );
+  private generateExecutables(
+    alerts: Record<string, Alert<State, Context, ActionGroupIds | RecoveryActionGroupId>>
+  ) {
+    const executables = [];
+
+    for (const action of this.rule.actions) {
+      if (isSummaryAction(action)) {
+        if (
+          !isSummaryActionThrottled({
+            action,
+            summaryActions: this.taskInstance.state?.summaryActions,
+            logger: this.logger,
+          })
+        ) {
+          executables.push({ action });
+        }
+        continue;
+      }
+      for (const [alertId, alert] of Object.entries(alerts)) {
+        const actionGroup = this.getActionGroup(alert);
+
+        if (!this.ruleTypeActionGroups!.has(actionGroup)) {
+          this.logger.error(
+            `Invalid action group "${actionGroup}" for rule "${this.ruleType.id}".`
+          );
+          continue;
+        }
+        if (action.group === actionGroup && !this.isAlertMuted(alertId)) {
+          if (
+            this.isRecoveredAlert(action.group) ||
+            this.isExecutableActiveAlert({ alert, action })
+          ) {
+            executables.push({ action, alert });
+          }
+        }
       }
-      this.skippedAlerts[alertId] = { reason: Reasons.MUTED };
-      return true;
     }
-    return false;
-  }
 
-  private isExecutableAction(action: RuleAction) {
-    return this.taskRunnerContext.actionsPlugin.isActionExecutable(action.id, action.actionTypeId, {
-      notifyUsage: true,
-    });
+    return executables;
   }
 
-  private isExecutableActiveAlert({
-    alert,
+  private async getSummarizedAlerts({
     action,
+    ruleId,
+    spaceId,
   }: {
-    alert: Alert<AlertInstanceState, AlertInstanceContext, ActionGroupIds | RecoveryActionGroupId>;
     action: RuleAction;
+    ruleId: string;
+    spaceId: string;
   }) {
-    const alertId = alert.getId();
-    const { rule, ruleLabel, logger } = this;
-    const notifyWhen = action.frequency?.notifyWhen || rule.notifyWhen;
-
-    if (notifyWhen === 'onActionGroupChange' && !alert.scheduledActionGroupHasChanged()) {
-      if (
-        !this.skippedAlerts[alertId] ||
-        (this.skippedAlerts[alertId] &&
-          this.skippedAlerts[alertId].reason !== Reasons.ACTION_GROUP_NOT_CHANGED)
-      ) {
-        logger.debug(
-          `skipping scheduling of actions for '${alertId}' in rule ${ruleLabel}: alert is active but action group has not changed`
-        );
-      }
-      this.skippedAlerts[alertId] = { reason: Reasons.ACTION_GROUP_NOT_CHANGED };
-      return false;
+    let options;
+
+    if (hasActionFrequencyThrottled(action)) {
+      const throttleMills = parseDuration(action.frequency!.throttle!);
+      const start = new Date(Date.now() - throttleMills);
```

Comment:
Wondering if we should use a more standardized time for the time range instead of Date.now()? Maybe task.startedAt? I think there are multiple rule types that link back to different views in the UI and being able to calculate this time range in a standardized way would be useful, either that or be able to provide these time bounds somehow so they can use it when building their links? cc @mikecote

Reply:
Good catch, we've had this oddity for a while as the same logic is used on the throttling per-alert feature. See https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/server/alert/alert.ts#L85-L89.

Some notes:
- What's there today sort of works when a rule run takes a while and makes Date.now() pass the throttle window. We'd be sending a notification now rather than the next rule run.
- Using task.startedAt would be more consistent when we use a precise scheduler to ensure actions send out consistently (ex: throttle of 5m would send notifications every 5m instead of 5m20s, 5m10s, etc like it sometimes does now).

Would be worth gathering people's thoughts, maybe as a discuss issue?

Reply:
Created a discuss issue: https://github.com/elastic/kibana/issues/147563

Reply:
Thank you!